Automattic/wp-calypso
# Hosting onboarding: Add flag for forcing hosting trial availability

Related to pdKhl6-2SQ-p2#comment-4897

## Proposed Changes

* Adds `hosting-trial-campaign` value to Calypso preferences store
* The `?campaign=reddit` query param will set `hosting-trial-campaign === 'reddit'`
* The flag in the onboarding store forces the free trial option to be shown, and prevents the user from being assigned to a test cohort if they haven't been already.

Implementation notes for @zaguiini 

I'm storing the `hosting-trial-campaign` flag in the store as an enum rather than a simple boolean because I have a hunch we might need to expand it in the future. We might want to be able to tell the difference between the flag being explicitly disabled vs. not set, or we might want to have `available-v1`, `available-v2`, etc. for different campaigns. We don't need the actions and selectors to reflect that, but I wanted the underlying storage mechanism to be more generic than a boolean.

~It seems to work, but I think the hackiest thing about the implementation is that the onboarding store gets imported anywhere the plans grid is rendered. e.g. if you go to `/plans/{{ site slug }}` the `useFreeHostingTrialAssignment` hook gets loaded, which uses the onboarding store. Just something to keep in mind, although it does appear to be fine.~

The flag is now stored as a user preference, making it sticky to the user and also meaning we don't need to import the onboarding store into the plans page.

## Testing Instructions



Before testing each case, remove the `explat-experiment--wpcom_hosting_business_plan_free_trial_v2` value from your `localStorage`. This will force Explat to fetch your assignment, meaning we can test whether the query param correctly prevents cohort assignment.

* Test cases where experiment assignment shouldn't happen
   * `/setup/new-hosted-site?campaign=reddit`
   * `/setup/new-hosted-site?flags=plans/hosting-trial`
   * `/plans/{{ site slug }}` (for some existing site)
   * `/start` (check the plans grid in this flow works as expected)
* Test cases where experiment assignment should happen
   * `/setup/new-hosted-site`

## Pre-merge Checklist
